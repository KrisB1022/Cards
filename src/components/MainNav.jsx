import React, { PureComponent } from "react";
import PropTypes from "prop-types";

import { Badge, Button, Navbar, NavbarBrand } from "reactstrap";
import logo from "../images/magic-banner-730x280.jpg";

class MainNav extends PureComponent {
  state = {
    showUserCards: false
  };

  showUserCards = () => {
    const { userCards, getCards } = this.props;

    getCards({
      clearCards: true,
      query: {
        id: Object.keys(userCards.byId)
      }
    });
    this.toggleUserCards();
  };

  hideUserCards = () => {
    this.props.applyFilters();
    this.toggleUserCards();
  };

  toggleUserCards = () =>
    this.setState(prevState => ({
      showUserCards: !prevState.showUserCards
    }));

  render() {
    const { showUserCards } = this.state;
    const { userCards, totalCount } = this.props;
    const numCards = Object.keys(userCards.byId).length;
    const hasNoCards = numCards === 0;

    return (
      <Navbar color="light" light expand="md" className="mb-3 sticky-top">
        <div className="container">
          <NavbarBrand href="/">
            <img
              className="img-fluid"
              src={logo}
              alt="Magic the Gathering logo"
              width={175}
            />
          </NavbarBrand>

          <p className="m-0">{`${totalCount} Cards`}</p>

          {showUserCards ? (
            <Button color="secondary" onClick={this.hideUserCards}>
              Hide my cards <Badge color="light">{numCards}</Badge>
            </Button>
          ) : (
            <Button
              color="primary"
              onClick={this.showUserCards}
              disabled={hasNoCards}
            >
              Show my cards <Badge color="light">{numCards}</Badge>
            </Button>
          )}
        </div>
      </Navbar>
    );
  }
}

MainNav.propTypes = {
  totalCount: PropTypes.number.isRequired,
  userCards: PropTypes.shape({
    byId: PropTypes.object.isRequired
  }).isRequired
};

export default MainNav;
