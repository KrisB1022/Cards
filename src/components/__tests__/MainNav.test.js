import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

import { Button, Navbar, NavbarBrand } from "reactstrap";
import MainNav from "../MainNav";
import logo from "../../images/magic-banner-730x280.jpg";

const totalCountMock = 20;
const userCardsMock = {
  byId: {}
};

let wrapper;

beforeEach(() => {
  wrapper = shallow(
    <MainNav totalCount={totalCountMock} userCards={userCardsMock} />
  );
});

it("renders without crashing", () => {
  shallow(<MainNav totalCount={totalCountMock} userCards={userCardsMock} />);
});

it("has 1 <NavBar/> and <NavBarBrand/>", () => {
  expect(wrapper.find(Navbar).exists()).toBeTruthy();
  expect(wrapper.find(NavbarBrand).exists()).toBeTruthy();
});

it("has 1 img", () => {
  expect(wrapper.find("img").exists()).toBeTruthy();
  expect(wrapper.find("img").prop("src")).toEqual(logo);
});

it("has Hide button if showUserCards = true", () => {
  wrapper.setState({
    showUserCards: true
  });

  const btn = wrapper.find(Button);

  expect(btn.exists()).toBeTruthy();
  expect(btn.prop("children")[0]).toMatch(/^Hide my cards/);
});

it("has Show button if showUserCards = false", () => {
  wrapper.setState({
    showUserCards: false
  });

  const btn = wrapper.find(Button);

  expect(btn.exists()).toBeTruthy();
  expect(btn.prop("children")[0]).toMatch(/^Show my cards/);
});

it("calls getCards and toggleUserCards for showUserCards", () => {
  const byId = { key1: true, key2: false };
  const getCardsMock = jest.fn();
  const spy = jest.spyOn(wrapper.instance(), "toggleUserCards");
  wrapper.setProps({
    userCards: { byId },
    getCards: getCardsMock
  });

  wrapper.instance().showUserCards();

  expect(getCardsMock).toHaveBeenCalledTimes(1);
  expect(getCardsMock).toHaveBeenCalledWith({
    clearCards: true,
    query: {
      id: Object.keys(byId)
    }
  });
  expect(spy).toHaveBeenCalledTimes(1);
});

it("calls applyFilters and toggleUserCards for hideUserCards", () => {
  const applyFiltersMock = jest.fn();
  const spy = jest.spyOn(wrapper.instance(), "toggleUserCards");
  wrapper.setProps({
    applyFilters: applyFiltersMock
  });

  wrapper.instance().hideUserCards();

  expect(applyFiltersMock).toHaveBeenCalledTimes(1);
  expect(spy).toHaveBeenCalledTimes(1);
});

it("toggles showUserCards for toggleUserCards", () => {
  wrapper.setState({
    showUserCards: false
  });

  wrapper.instance().toggleUserCards();
  expect(wrapper.state().showUserCards).toBeTruthy();

  wrapper.instance().toggleUserCards();
  expect(wrapper.state().showUserCards).toBeFalsy();
});

it("matches snapshot", () => {
  const tree = renderer.create(
    <MainNav totalCount={totalCountMock} userCards={userCardsMock} />
  );

  expect(tree.toJSON()).toMatchSnapshot();
});
