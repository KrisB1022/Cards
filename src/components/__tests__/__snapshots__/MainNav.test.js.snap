// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<nav
  className="mb-3 sticky-top navbar navbar-expand-md navbar-light bg-light"
>
  <div
    className="container"
  >
    <a
      className="navbar-brand"
      href="/"
    >
      <img
        alt="Magic the Gathering logo"
        className="img-fluid"
        src="magic-banner-730x280.jpg"
        width={175}
      />
    </a>
    <p
      className="m-0"
    >
      20 Cards
    </p>
    <button
      aria-label={null}
      className="btn btn-primary disabled"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Show my cards 
      <span
        className="badge badge-light"
      >
        0
      </span>
    </button>
  </div>
</nav>
`;
